lib/gameState.ex

```diff
@@ -0,0 +1,98 @@
+defmodule GameState do
+  def new_game_state do
+    %{
+      :available_moves => [1, 2, 3, 4, 5, 6, 7, 8, 9],
+      :win? => false,
+      :player_1_moves => [],
+      :player_2_moves => [],
+      :player_1_mark => "X",
+      :player_2_mark => "O",
+      :current_player => 1
+    }
+  end
+
+  def win?(game_state) do
+    game_state[:win?]
+  end
+
+  def set_player_1_mark(new_marker, game_state) do
+    %{game_state | player_1_mark: new_marker}
+  end
+
+  def set_player_2_mark(new_marker, game_state) do
+    %{game_state | player_2_mark: new_marker}
+  end
+
+  def set_game_state_win(win_status, game_state) do
+    %{game_state | win?: win_status}
+  end
+
+  def check_for_win(game_state) do
+    move_list = determine_move_list(game_state)
+
+    win_status =
+      game_state[move_list]
+      |> check_for_win_combos()
+
+    set_game_state_win(win_status, game_state)
+  end
+
+  def check_for_win_combos(moves_list) do
+    win_combos = [
+      [1, 2, 3],
+      [4, 5, 6],
+      [7, 8, 9],
+      [1, 4, 7],
+      [2, 5, 8],
+      [3, 6, 9],
+      [1, 5, 9],
+      [3, 5, 7]
+    ]
+
+    win_combo_list =
+      Enum.map(win_combos, fn combo ->
+        Enum.map(combo, fn num ->
+          num in moves_list
+        end)
+      end)
+
+    win_list =
+      Enum.map(win_combo_list, fn combo ->
+        Enum.all?(combo)
+      end)
+
+    Enum.any?(win_list)
+  end
+
+  def change_player(game_state) do
+    %{game_state | current_player: (if get_current_player(game_state) == 1, do: 2, else: 1)}
```

Comment:
The get_current player is an abstraction that I wanted to keep around. Just in case I change how the data structure is called on, this gives me the ability to make my change in one location for naming the key in the game_state. I also have to get the current player a few times, so having it abstracted out for someone who is new to ruby should help with people looking at the code in the future.

As for pattern matching, I'll try to look into how to use that here.

Reply:
Sure here's two different ways–

**w/o a guard clause**
```elixir
def change_player(%{current_player: 1} = game), do: %{game | current_player: 2}
def change_player(game), do: %{game | current_player: 1}
```

**using guard clause**
```elixir
def change_player(game) when game.current_player == 1, do: %{game | current_player: 2}
def change_player(game), do: %{ game | current_player: 1}
```